### PR TITLE
Add api traits

### LINKED ISSUE #267: Abstract api-client main features behind a trait
This would greatly simplify the separation between the core logic and the custom business logic in the integritee-worker in case the node has some customizations, which are incompatible with the default api-client type, e.g., the signed extra if  `pallet_asset_tx_payment` is used.

Excerpt from the envisioned trait

```rust

pub trait Api {
          fn get_storage<V>() -> Result<Option<V>>;
     
          fn subscribe_events();

        // etc.
}
```

### PR DESCRIPTION
**Pull request by haerdib**
Adds traits to the api and removes obsolete impl enforcements on the Api where possible (see #357)

closes #267

### CODE REVIEW

File: src/api/error.rs [hunk 1]
```diff
@@ -33,15 +33,14 @@ pub enum Error {
 	MetadataFetch,
 	#[error("Operation needs a signer to be set in the api")]
 	NoSigner,
-	#[error("RpcClient error: {0:?}")]
+	#[error("RpcClient  error: {0:?}")]
```
Comment (clangenb) [inline src/api/error.rs:36]: typo

File: src/api/interfaces/frame_system.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
+/*
```
Comment (clangenb) [inline src/api/interfaces/frame_system.rs:1]: Minor naming convention: in rust you talk about traits, not interfaces.

Second, with this name, I would expect only the trait declarations, not the implementations here.

File: src/api/interfaces/generic_storage.rs [hunk 3]
```diff
@@ -0,0 +1,255 @@
+/*
```
Comment (clangenb) [inline src/api/interfaces/generic_storage.rs:1]: I would just call it `storage` the name already implies generic.

File: src/api/interfaces/frame_system.rs (hunk 2 above)
Comment (clangenb) [reply src/api/interfaces/frame_system.rs:1]: Maybe call it `rpc_implementations`?

File: src/api/interfaces/pallet_balances.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
+/*
```
Comment (clangenb) [inline src/api/interfaces/pallet_balances.rs:1]: Hmm, `pallet_balances` is wrong actually, only the existential deposit belongs to the pallet balances. The rest is `pallet-transaction-payment`

File: src/api/interfaces/frame_system.rs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	Runtime: FrameSystemConfig,
+	Params: ExtrinsicParams<Runtime::Index, Runtime::Hash>,
+	Runtime::Header: DeserializeOwned,
+	Runtime::Hash: FromHexString,
+{
+	type Header = Runtime::Header;
+
+	fn get_finalized_head(&self) -> ApiResult<Option<Runtime::Hash>> {
+		let finalized_block_hash =
+			self.client().request("chain_getFinalizedHead", rpc_params![])?;
```
Comment (clangenb) [inline src/api/interfaces/frame_system.rs:115]: this is obviously not from the `frame-system-pallet`

**Review by clangenb — COMMENTED** [review]
Finally, the monolith has been separated!

However, I think that the separation is not really coherent with the substrate structure. What I propose to cleanly separate the stuff:

```
rpc_implementatsions/
   chain
   state
   constants
   frame_system
   pallet-balances
   pallet-transaction-payment
```
I do wonder if we should split up the subscriptions and put them into the respective domain, e.g.,

`subscribe_finalized_head` -> `chain` is this queries the `chain` RPC of the substrate node.

On the other hand, this would introduce a bit more complexity. I am unsure.

File: src/api/interfaces/frame_system.rs (hunk 2 above)
Comment (haerdib) [reply src/api/interfaces/frame_system.rs:1]: I was not thinking about rust interfaces (e.g. traits) but rather the node interface. But, agreed, it is confusing.

Comment (haerdib) [reply src/api/interfaces/frame_system.rs:1]: How about `rpc_interfaces`? It's not really an implementation, I think. Only a helper for easy rpc access and forwarding it to the client.

Comment (haerdib) [reply src/api/interfaces/frame_system.rs:1]: Or `rpc_helpers`?

File: src/api/interfaces/generic_storage.rs (hunk 3 above)
Comment (haerdib) [reply src/api/interfaces/generic_storage.rs]: Renamed to `state`

File: src/api/interfaces/pallet_balances.rs (hunk 4 above)
Comment (haerdib) [reply src/api/interfaces/pallet_balances.rs:1]: You're right, moved :)

File: src/api/interfaces/frame_system.rs (hunk 5 above)
Comment (haerdib) [reply src/api/interfaces/frame_system.rs]: Moved to chain.rs

File: src/api/interfaces/subscription.rs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+};
+pub use ac_node_api::{events::EventDetails, StaticEvent};
+use ac_node_api::{DispatchError, Events};
+use ac_primitives::{ExtrinsicParams, FrameSystemConfig};
+use log::*;
+use serde::de::DeserializeOwned;
+use sp_core::storage::StorageChangeSet;
+
+// FIXME: This should rather be implemented directly on the
+// Subscription return value, rather than the api. Or directly
```
Comment (haerdib) [inline src/api/interfaces/subscription.rs:26]: I don't like how it's currently done. But changing this would break the interface. So that should be done in a different PR

**Review by haerdib — COMMENTED** [review]
Moved all subscription, where possible, to the correspondent file. I think it makes more sense than to have all subscriptions in one file.

File: src/api/interfaces/generic_storage.rs (hunk 3 above)
Comment (clangenb) [reply src/api/interfaces/generic_storage.rs]: Thanks!

File: examples/event_callback.rs [hunk 7]
```diff
@@ -38,7 +38,7 @@ async fn main() {
 		Api::<sr25519::Pair, _, AssetTipExtrinsicParams<Runtime>, Runtime>::new(client).unwrap();
 
 	println!("Subscribe to events");
-	let mut subscription = api.subscribe_events().unwrap();
+	let mut subscription = api.subscribe_system_events().unwrap();
```
Comment (clangenb) [inline examples/event_callback.rs:41]: Good renaming!

File: src/api/interfaces/pallet_balances.rs (hunk 4 above)
Comment (clangenb) [reply src/api/interfaces/pallet_balances.rs:1]: Yaay! 🥳

File: src/api/interfaces/state.rs [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+use ac_compose_macros::rpc_params;
+pub use ac_node_api::{events::EventDetails, StaticEvent};
+use ac_primitives::{ExtrinsicParams, FrameSystemConfig};
+use codec::{Decode, Encode};
+use log::*;
+use serde::de::DeserializeOwned;
+use sp_core::storage::{StorageChangeSet, StorageData, StorageKey};
+
+/// Generic interface to substrate storage.
+pub trait GetState<Hash> {
```
Comment (clangenb) [inline src/api/interfaces/state.rs:23]: I would change this to `GetStorage`

File: src/api/interfaces/frame_system.rs (hunk 2 above)
Comment (clangenb) [reply src/api/interfaces/frame_system.rs:1]: I would still vote for `rpc_implementations` they are not helpers for the RPC, they are actual (client) implementations in my opinion. Or maybe `rpc-api`?

File: src/api/interfaces/subscription.rs (hunk 6 above)
Comment (clangenb) [reply src/api/interfaces/subscription.rs:26]: Now, we only have events here. Maybe we can call it `EventSubscription`?

File: src/api/interfaces/subscription.rs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+// subcribe. Should be looked at in #288
+// https://github.com/scs/substrate-api-client/issues/288#issuecomment-1346221653
+pub trait EventSubscriptionHelper<Client, Hash>
```
Comment (clangenb) [inline src/api/interfaces/subscription.rs:29]: I would drop the helper here.

File: src/api/interfaces/state.rs (hunk 8 above)
Comment (haerdib) [reply src/api/interfaces/state.rs:23]: There's already a GetStorage in node_api to retrieve metadata storage values. I'd rather avoid having two equally named storage.

Two options, if you don't like `GetState`:
- rename the `GetStorage` from node api to `GetStorageTypes`
- rename `GetState` to `GetStateStorage` 

or similar naming.. what do you think?

File: src/api/interfaces/subscription.rs (hunk 6 above)
Comment (haerdib) [reply src/api/interfaces/subscription.rs:26]: Ah, actually, following the rust convention, that should rather be a verb noun. How about `SubscribeEvents`?

File: src/api/interfaces/frame_system.rs (hunk 2 above)
Comment (haerdib) [reply src/api/interfaces/frame_system.rs:1]: 👍 rpc-api

File: src/api/interfaces/state.rs (hunk 8 above)
Comment (haerdib) [reply src/api/interfaces/state.rs]: Went for the first one now - I think `GetStorageTypes` makes more sense anyhow.

Comment (clangenb) [reply src/api/interfaces/state.rs]: Yes, I agree! 👍

File: src/api/interfaces/subscription.rs (hunk 6 above)
Comment (clangenb) [reply src/api/interfaces/subscription.rs]: Yes, I agree that is even better, but the last thing would to also rename the file accordingly to `subscribe_events` :)

Comment (haerdib) [reply src/api/interfaces/subscription.rs]: Done :)

**Review by clangenb — APPROVED** [review]
Whoopwhoop!
